localstack/aws/protocol/router.py

```diff
@@ -0,0 +1,234 @@
+import re
+from collections import defaultdict
+from typing import Any, Dict, List, Mapping, NamedTuple, Optional, Tuple
+from urllib.parse import parse_qs, unquote
+
+from botocore.model import OperationModel, ServiceModel, StructureShape
+from werkzeug.datastructures import Headers, MultiDict
+from werkzeug.exceptions import NotFound
+from werkzeug.routing import Map, MapAdapter, Rule
+
+from localstack.http import Request
+from localstack.http.request import get_raw_path
+
+
+class _RequiredArgsRule:
+    """
+    Specific Rule implementation which checks if a set of certain required header and query parameters are matched by
+    a specific request.
+    """
+
+    endpoint: Any
+    required_query_args: Optional[Mapping[str, List[Any]]]
+    required_header_args: List[str]
+    match_score: int
+
+    def __init__(
+        self,
+        endpoint: Any,
+        required_query_args: Optional[Mapping[str, List[Any]]],
+        required_header_args: Optional[List[str]],
+    ) -> None:
+        super().__init__()
+        self.endpoint = endpoint
+        self.required_query_args = required_query_args or {}
+        self.required_header_args = required_header_args or []
+        self.match_score = 1 + len(self.required_query_args) + len(self.required_header_args)
+
+    def matches(self, query_args: "MultiDict[str, str]", headers: Headers) -> bool:
```

Comment:
any reason we are using the double quotes around the `MultiDict` type hint here?